include/llpc.h

```diff
@@ -239,16 +239,80 @@ struct ShaderModuleBuildInfo
 #endif
 };
 
-/// Represents the header part of LLPC shader module data
-struct ShaderModuleDataHeader
+/// Represents the base data type
+enum class BasicType : uint32_t
 {
-    uint32_t hash[4];       // Shader hash code
+    Unknown = 0,          ///< Unknown
+    Float,                ///< Float
+    Double,               ///< Double
+    Int,                  ///< Signed integer
+    Uint,                 ///< Unsigned integer
+    Int64,                ///< 64-bit signed integer
+    Uint64,               ///< 64-bit unsigned integer
+    Float16,              ///< 16-bit floating-point
+    Int16,                ///< 16-bit signed integer
+    Uint16,               ///< 16-bit unsigned integer
+    Int8,                 ///< 8-bit signed integer
+    Uint8,                ///< 8-bit unsigned integer
+};
+
+/// Enumerates types of shader binary.
+enum class BinaryType : uint32_t
+{
+    Unknown = 0,  ///< Invalid type
+    Spirv,        ///< SPIR-V binary
+    LlvmBc,       ///< LLVM bitcode
+    MultiLlvmBc,  ///< Multiple LLVM bitcode
+    Elf,          ///< ELF
+};
+
+/// Represents the information of one shader entry in ShaderModuleExtraData
+struct ShaderModuleEntryData
+{
+    ShaderStage stage;              ///< Shader stage
+    void*       pShaderEntry;       ///< Private shader module entry info
+};
+
+/// Represents usage info of a shader module
+struct ShaderModuleUsage
+{
+    bool                  enableVarPtrStorageBuf;  ///< Whether to enable "VariablePointerStorageBuffer" capability
+    bool                  enableVarPtr;            ///< Whether to enable "VariablePointer" capability
+    bool                  useSubgroupSize;         ///< Whether gl_SubgroupSize is used
+    bool                  useHelpInvocation;       ///< Whether fragment shader has helper-invocation for subgroup
+    bool                  useSpecConstant;         ///< Whether specializaton constant is used
+    bool                  keepUnusedFunctions;     ///< Whether to keep unused function
+};
+
+/// Represents common part of shader module data
+struct ShaderModuleData
+{
+    uint32_t         hash[4];       ///< Shader hash code
+    BinaryType       binType;       ///< Shader binary type
+    BinaryData       binCode;       ///< Shader binary data
+    uint32_t         cacheHash[4];  ///< Hash code for calculate pipeline cache key
+    ShaderModuleUsage usage;         ///< Usage info of a shader module
+};
+
+/// Represents extra part of shader module data
+struct ShaderModuleExtraData
+{
+    uint32_t              entryCount;              ///< Shader entry count in the module
+    ShaderModuleEntryData entryDatas[1];           ///< Array of all shader entries in this module
+};
+
+
+/// Represents extended output of building a shader module (taking extra data info)
+struct ShaderModuleDataEx
+{
+    ShaderModuleData        common;        ///< Shader module common data
+    ShaderModuleExtraData   extra;         ///< Shader module extra data
```

Comment:
So if I'm understanding this right, the intention here is for `ShaderModuleExtraData::entryDatas` to be effectively a variable-sized array, which contains multiple array entries, one per entrypoint function in the module.

That seems fine, except by embedding `ShaderModuleExtraData` into `ShaderModuleDataEx`, we're now in the situation where `ShaderModuleDataEx` has a variable-length array embedded inside of it, but not at the top-level -- which is a risk to introducing bugs.

Since the `ShaderModuleExtraData` structure doesn't appear to be used standalone anywhere else, could you just inline its members into `ShaderModuleDataEx`? In that way, it is clear that there is a variable-length array at the end.

Reply:
good idea. Will update it.

Reply:
Thank you, LGTM